### PR TITLE
Raise user-friendly ex when aad auth failed.

### PR DESCRIPTION
**Pull request by terencefan**
According to [this](https://docs.microsoft.com/en-us/azure/role-based-access-control/troubleshooting#role-assignment-changes-are-not-being-detected), a role assignments change requires up to 30 minutes to take effect if the token had not been forced refreshed. 

However, it took seldomly longer than 15 minutes in our daily integration tests.

I'd like to try reducing the time to take effect, but now, let's give our customer some hints.

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs [hunk 1]
```diff
@@ -9,12 +9,16 @@ namespace Microsoft.Azure.SignalR.Common
     [Serializable]
     public class AzureSignalRUnauthorizedException : AzureSignalRException
     {
-        private const string ErrorMessage = "Authorization failed. Make sure you provide the correct connection string and have access to the resource.";
+        private const string ErrorMessage = "Authorization failed. Please check the connection string and role assignments. The role assignments will take up to 30 minutes to take effect if it was added recently.";
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs:12]: use different error messages for accessKey auth and aad auth?

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerImpl.cs [hunk 2]
```diff
@@ -7,10 +7,8 @@
 using System.Security.Claims;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.Azure.SignalR.Common.RestClients;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Management/ServiceManagerImpl.cs:10]: ad a general practice, if the file only changes format, revert the change to keep the code change PR minimum. (Format change can be in a separate PR)

File: src/Microsoft.Azure.SignalR.Common/Exceptions/ExceptionExtensions.cs [hunk 3]
```diff
@@ -35,5 +37,19 @@ internal static Exception WrapAsAzureSignalRException(this Exception e, Uri base
                     return e;
             }
         }
+
+        internal static Exception WrapAsAzureSignalRException(this Exception e, WebSocketConnectionOptions options)
+        {
+            switch (e)
+            {
+                case WebSocketException webSocketException:
+                    if (options.AuthType == AuthType.AzureAd && e.Message.Contains("401"))
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/ExceptionExtensions.cs:46]: looks cool, can it be more specific using "StartWith"? Could it be also applied to other auth types?

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs (hunk 1 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs:12]: This message is only for AuthType = AzureAd

File: src/Microsoft.Azure.SignalR.Management/ServiceManagerImpl.cs (hunk 2 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Management/ServiceManagerImpl.cs:10]: I relocate this file and the namespace is no longer exists, so I think this change relates to the PR.

File: src/Microsoft.Azure.SignalR.Common/Exceptions/ExceptionExtensions.cs (hunk 3 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Exceptions/ExceptionExtensions.cs:46]: 401 makes the most sense in that error message. If we use the `StartWith`, it could be `.StartWith("The server returned status code '401'"), which is no better than `Contains`.

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs (hunk 1 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs:12]: The exception class name looks much more general though.

File: src/Microsoft.Azure.SignalR.Common/Exceptions/ExceptionExtensions.cs (hunk 3 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Exceptions/ExceptionExtensions.cs:46]: Is it possible that other place contains some random 401 number that is not related to status code?

Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Exceptions/ExceptionExtensions.cs:46]: The chance is low. They doesn't contain random strings as I know.

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs (hunk 1 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs:12]: It's a public class, it doesn't make sense if we raised an exception with its type name contains "AzureAd".

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs [hunk 4]
```diff
@@ -9,12 +9,16 @@ namespace Microsoft.Azure.SignalR.Common
...
 
         public AzureSignalRUnauthorizedException(string requestUri, Exception innerException) : base(string.IsNullOrEmpty(requestUri) ? ErrorMessage : $"{ErrorMessage} Request Uri: {requestUri}", innerException)
         {
         }
 
+        internal AzureSignalRUnauthorizedException(Exception innerException) : base(ErrorMessage, innerException)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs:18]: Let's either use a more specific Aad exception or pass in the AAd exception message, instead of having aad specific error message inside this general exception class

File: src/Microsoft.Azure.SignalR.Common/Exceptions/ExceptionExtensions.cs [hunk 5]
```diff
@@ -35,5 +37,19 @@ internal static Exception WrapAsAzureSignalRException(this Exception e, Uri base
                     return e;
             }
         }
+
+        internal static Exception WrapAsAzureSignalRException(this Exception e, WebSocketConnectionOptions options)
+        {
+            switch (e)
+            {
+                case WebSocketException webSocketException:
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Exceptions/ExceptionExtensions.cs:45]: let's use `StartWith` instead of `Contains` to make the check more straight-forward

File: src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs (hunk 4 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Exceptions/AzureSignalRUnauthorizedException.cs:18]: It's impossible.

First, we can't change an existing public exception name.
Second, it's impossible to determine if the a request was authorized by `AAD Token` or `AccessToken` when calling rest APIs.

**Review by vicancy — APPROVED** [review]
